Sources/SwiftParserDiagnostics/MissingNodesError.swift

```diff
@@ -36,7 +36,11 @@ fileprivate enum NodesDescriptionPart {
     switch self {
     case .tokensWithDefaultText(var tokens):
       if format {
-        tokens = tokens.map({ NoNewlinesFormat(viewMode: .all).visit($0) })
+        tokens = tokens.enumerated()
+          .map { index, token in
+            // We detach the nodes from the second one onwards, considering the line indentation only for the first node
+            NoNewlinesFormat(viewMode: .all).visit(index == 0 ? token : token.detached)
+          }
```

Comment:
Why is this needed? I would expect that we should be able to always do

```diff
- tokens = tokens.map({ NoNewlinesFormat(viewMode: .all).visit($0) })
+ tokens = tokens.map({ NoNewlinesFormat(viewMode: .all).visit($0.detached) })
```

Reply:
I seem to have misunderstood, and now I agree with you.

However, [this]([cd3ad12](https://github.com/apple/swift-syntax/pull/1867/commits/cd3ad12ce53e826b6583e609528e7ac5b1994e7e)) will cause a test to fail for this test case.

https://github.com/apple/swift-syntax/blob/e7068e8eb76542591a189f703f8b62881a520588/Tests/SwiftParserTest/AttributeTests.swift#L99-L113


The error message appears in this line of the DiagnosticSpec, stating that it outputs `:false` instead of `: false`.
https://github.com/apple/swift-syntax/blob/e7068e8eb76542591a189f703f8b62881a520588/Tests/SwiftParserTest/AttributeTests.swift#L106

The whitespace between `:` and `false` should be inserted by BasicFormat, and I believe the code responsible for it is [here](https://github.com/apple/swift-syntax/blob/e7068e8eb76542591a189f703f8b62881a520588/Sources/SwiftBasicFormat/BasicFormat.swift#L416-L425).

But `requiresWhitespace(between:and:)` returns false because the detached `:` token will not find `false` as its next token. As a result, the whitespace is not inserted as expected in the test case.

I am currently investigating this issue.

Reply:
Ah, I see what’s going wrong g here: We definitely shouldn’t `detach` because otherwise we don’t know that `:` is followed by `false` in `NoNewlinesFormat` and thus can’t decide whether to add a space after it. 

The issue is that inheriting the line’s indentation only makes sense if the formatted text should contain newlines. `NoNewlinesFormat` is a bit of an oddity here because it doesn’t do that. What I would do is to add the following property to `BasicFormat` and only perform your logic if `inferInitialTokenIndentaiton` is `true`. `NoNewlinesFormat` can then override it and set it to `false`.

```swift
/// If we are formatting a subtree and the line that the initial token occurs on is indented,
/// use that line indentation for the first token in the subtree to format.
/// 
/// For example, when formatting only the code block in the following,
/// then the opening `{` should be indented by four spaces.
/// ```
///     func test() {
///         print(1)
///     }
/// ```
open var inferInitialTokenIndentaiton: Bool { true }
```

And then, I think we also don’t need the inline comment in `visit(_: TokenSyntax)` anymore because the behavior is documented on `inferInitialTokenIndentaiton`.

Reply:
Wow, thank you for your great support!
I fixed it and squashed those commits into cc9f4496099a0596dcf99da6881551aa4e590c78